Removing the e2e functionality for v4 cleanup

azoppiserpa opened the pull request:
### Which issue this PR addresses:

<!--
Please include a link to the ADO work item as well as any GitHub issues.

Usage: `Fixes #<GitHub issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes JIRA: https://issues.redhat.com/browse/ARO-4222

### What this PR does / why we need it:

<!--
Include a brief summary of what the PR is intended to accomplish and how the PR
does it. (2-3 sentences)
-->

The PR intends to retire billing v4 E2E. This is done by removing the function `createOrUpdateE2EBlob` and its usages. Related environment variables, feature flags and other elements were also investigated and removed.

### Test plan for issue:

<!--
How did you test that this PR works?

- Are there unit tests?
- Are there integration/e2e tests?
- If it is not possible to write automated tests, explain why and document how
  to manually test and verify the feature.
-->

Unit tests -> `go test` on each related subdirectory of `pkg` passed.
Integration/e2e/other -> E2E test was run [here](https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=91369638&view=logs&j=4c16f503-2748-5cc0-5ca9-498b68f4d56e&t=bc7de246-111d-46ca-8cd6-7a3e4251458b&s=a9383a0e-f961-5bbc-dd7a-2d947d87ad0d). It failed, but the failure is due to a known bug during the cleanup after the E2E tests have already finished and passed themselves.

### Is there any documentation that needs to be updated for this PR?

<!--
- If yes and the docs are in GitHub, include doc updates in the PR.
- If yes and the docs are not in GitHub (i.e. ADO wiki), include a link to the
  docs.
- If no, explain why (e.g. "tech debt cleanup, N/A").
-->

Yes, the README from `RP-Config`, because we need to remove the E2E storage account parameter, and it is listed on the README. It has been updated on a different PR.

### How do you know this will function as expected in production? 

<!--
- Does adequate telemetry, monitoring and documentation exist to effectively operate your change?
- Have failure modes been identified and mitigated? 
-->

I'm retiring code to remove functionality. I believe we will know if anything I've removed causes a breaking change during build or the deploy attempt, and it shouldn't change the behavior of anything in production. Please let me know if I am wrong so that I can validate it better :)

mociarain commented:
/azp run ci

mociarain reviewed (commented):
There seem to be a few other things in this file (and other places) that might warrant cleaning up as part of this PR:
- This [function](https://github.com/azure/ARO-RP/blob/e5c40654277c77fe78ba669610ac05774e448683/docs/feature-flags.md#L23) is no longer being used.
- We are still referencing E2E stuff in the [storage client](https://github.com/azure/ARO-RP/blob/0534b0e7b0979371c9a14428afd445db1801d4fd/pkg/util/billing/billing.go#L55)

Also I have a few questions:

- Can you link the other PR you mentioned?
- Are there plans to remove the [flag that fronts this feature](https://github.com/azure/ARO-RP/blob/e5c40654277c77fe78ba669610ac05774e448683/docs/feature-flags.md#L23)?
- Should we remove this [env var](https://github.com/azure/ARO-RP/blob/0534b0e7b0979371c9a14428afd445db1801d4fd/pkg/deploy/generator/scripts/rpVMSS.sh#L268) reference: ?

kimorris27 reviewed (commented):
+1 to @mociarain's comments - I agree that it seems like we can get rid of the entire `storageClient` function and the env var it references (`BILLING_E2E_STORAGE_ACCOUNT_ID`).

azoppiserpa commented:
Reminder to myself - things to improve together with current comments:

Remove E2E from config object: https://github.com/Azure/ARO-RP/blob/5cd28ea9ed26a4c70a7cf628398803d345f41099/pkg/deploy/config.go#L51

azoppiserpa commented:
Reminder to myself - check where this folder is used (billing v4 only?), because maybe it can be deleted entirely: https://github.com/Azure/ARO-RP/tree/5cd28ea9ed26a4c70a7cf628398803d345f41099/pkg/util/billing

azoppiserpa commented:
@mociaran here's the previous PR link you requested: https://github.com/Azure/ARO-RP/pull/3484

notes about it: it does the same thing, but it wasn't following the new naming conventions for the branches. It was also not good to run E2E, because it had been created on a fork, so I moved it all to this current PR when @kimorris27 let me know I shouldn't be using the fork

regarding your other questions, I'm still investigating them, will update asap.

azoppiserpa commented:
**Note to reviewers:** did two commits but it's still a WIP, not a final version yet.

**What I did and why:**

I investigated the usages and removed the `storageClient` function, as mentioned by Kipp and Maitiú.

Regarding the environment variable, I removed it from the `rpVMSS.sh` script on a separate commit (for easy revertability if needed), as I am still not sure if we can't remove it entirely. I am looking up how it is defined, there are some related usages to it on `templates_rp.go` and `resources_rp.go`, which look like can be removed too. I am still testing a little bit locally to see if further removals break anything.

**What I am doing next:**
Evaluate the feature flags.
Finish evaluating env var and removing depending on evaluation result.

azoppiserpa commented:
Done on latest commits:
- Removed `billingE2EStorageAccountId` from variables lists, parameters lists and `Config` structs, after studying its usages and the flow of the generator code

Yet to do:
- Investigate removal from `assets/rp-production.json`, as its usage seems a little different than `assets/rp-production-parameters.json`
- Investigate `pkg/util/billing` folder retirement possibility
- Investigate feature flags

azoppiserpa commented:
Reminder to myself:
- The checker failed on `ci-go / generate-check` after my commit `removing env var part 1`, as well as on subsequent commits. I need to understand if my commit is wrong or if I need to update the checker later.

azoppiserpa commented:
**Update regarding the possibility of retiring the whole `util/billing` folder:** we can't do that, because `Delete` and `Ensure` methods populate a database that Billing V5 uses. There was an unused function we could remove though.

azoppiserpa commented:
INT run: https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=91369638&view=results

azoppiserpa commented:
@mociarain and @kimorris27 , I've removed the environment variable and the `storageClient` function. Will be investigating the feature flags right now.

azoppiserpa commented:
@mociarain removed the feature flags as requested.

Next steps: investigate why it failed on ev2.

mociarain commented:
/azp run ci

mociarain commented:
@azoppiserpa Thanks for the updates but don't feel obliged to give them. Tbh I just wait for the "re-review" ask but happy to give intermediate reviews if I can. Also feel free to ping me if there's something I can help with

whober0521 commented on pkg/deploy/config.go:
```diff
@@ -48,7 +48,6 @@ type Configuration struct {
 	ARMAPIClientCertCommonName         *string                `json:"armApiClientCertCommonName,omitempty"`
 	ARMClientID                        *string                `json:"armClientId,omitempty"`
-	BillingE2EStorageAccountID         *string                `json:"billingE2EStorageAccountId,omitempty"`
 	BillingServicePrincipalID          *string                `json:"billingServicePrincipalId,omitempty"`
```
where is BillingServicePrincipalID used?

azoppiserpa replied:
Indeed it is no longer being used and can also be removed. I investigated it last week.
I created a separate Jira issue for this, intending to remove it on a separate pull request.
Is it ok to remove `BillingServicePrincipalId` on a different pull request? 
Or would you prefer that I do it on this one?

whober0521 replied:
let's delete it in this PR

niontive replied:
To keep things simpler, I'm good with having one PR for billing e2e config and another PR for the unused billing service principal ID.

If you want to delete both in this PR, that's fine with me as well - I'll approve either.

niontive replied:
Just make sure it's clear in your PR description that you'll cleanup the other stuff in separate PRs.

azoppiserpa replied:
Removed `billingServicePrincipalId` on the new commits below. Also removed a role assignment that was being given to that `billingServicePrincipalId`, because it made sense to me that if we no longer have that service principal ID we no longer need to assign it any role. But is that conclusion correct? If the role assignment should be kept, please let me know.

niontive commented on docs/feature-flags.md:
```diff
@@ -20,10 +20,6 @@ Subscription feature flags used by ARO-RP include:
   graduating features to production such that they only apply on our Red Hat
   engineering subscription.
 
-* Microsoft.RedHatOpenShift/SaveAROTestConfig: used on our test subscriptions to
```
Make sure to also cleanup the following:
- https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/WIKI-Designs?path=/Build%252Dout/ARM-Deployment-and-Management.md&_a=preview&version=GBmaster
- https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/ARO-Pipelines?path=%2Farm%2Ffeature%2Fconfig.yaml&_a=contents&version=GBmaster

If I were you, I'd make a separate PR to remove `Microsoft.RedHatOpenShift/SaveAROTestConfig`

azoppiserpa replied:
Cleaned up the `config.yaml` on `ARO-Pipelines`. PR: https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/ARO-Pipelines/pullrequest/9912014

Also cleaned up the wiki.

whober0521 commented on pkg/util/billing/billing.go:
```diff
@@ -34,58 +21,21 @@ type Manager interface {
-	subDB         database.Subscriptions
-	log           *logrus.Entry
+	billingDB database.Billing
+	subDB     database.Subscriptions
```
where is subDB used?

azoppiserpa replied:
Removed :)

tsatam reviewed (commented):
Reviewed during sync session. The changes LGTM outside of the pending comments still remaining.

mociarain commented:
/azp run ci

mociarain reviewed (dismissed):
This looks great! So good seeing so much red ;)

mociarain commented:
/azp run ci

kimorris27 commented:
/azp run ci, e2e

cadenmarchese commented:
/azp run ci

azoppiserpa commented:
/azp run ci
